src/n_body.rs

```diff
@@ -1,144 +1,224 @@
+                0.,
+            ),
+            mass: 2.85885980666130812e-04 * SOLAR_MASS,
+        },
+        // uranus:
+        Body {
+            x: F64x4::new(
+                1.28943695621391310e+01,
+                -1.51111514016986312e+01,
+                -2.23307578892655734e-01,
+                0.,
+            ),
+            v: F64x4::new(
+                2.96460137564761618e-03 * DAYS_PER_YEAR,
+                2.37847173959480950e-03 * DAYS_PER_YEAR,
+                -2.96589568540237556e-05 * DAYS_PER_YEAR,
+                0.,
+            ),
+            mass: 4.36624404335156298e-05 * SOLAR_MASS,
+        },
+        // neptune:
+        Body {
+            x: F64x4::new(
+                1.53796971148509165e+01,
+                -2.59193146099879641e+01,
+                1.79258772950371181e-01,
+                0.,
+            ),
+            v: F64x4::new(
+                2.68067772490389322e-03 * DAYS_PER_YEAR,
+                1.62824170038242295e-03 * DAYS_PER_YEAR,
+                -9.51592254519715870e-05 * DAYS_PER_YEAR,
+                0.,
+            ),
+            mass: 5.15138902046611451e-05 * SOLAR_MASS,
+        },
+    ]
+}
+
+pub fn offset_momentum(bodies: &mut [Body; N_BODIES]) {
+    let (sun, rest) = bodies.split_at_mut(1);
+    let sun = &mut sun[0];
+    for body in rest {
+        let m_ratio = body.mass / SOLAR_MASS;
+        sun.v -= body.v * m_ratio;
     }
 }
 
-fn energy(bodies: &[Planet;N_BODIES]) -> f64 {
-    bodies.iter().enumerate().fold(0.0, |e, (i, bi)| {
-        let ei = (bi.vx * bi.vx + bi.vy * bi.vy + bi.vz * bi.vz) * bi.mass / 2.0;
-        let ei2 = bodies[i + 1..].iter().fold(0.0, |ej, bj| {
+pub fn energy(bodies: &[Body; N_BODIES]) -> f64 {
+    let mut e = 0.;
+    for i in 0..N_BODIES {
+        let bi = &bodies[i];
+        e += bi.mass * (bi.v * bi.v).sum() * 0.5;
+        for bj in bodies.iter().take(N_BODIES).skip(i + 1) {
             let dx = bi.x - bj.x;
-            let dy = bi.y - bj.y;
-            let dz = bi.z - bj.z;
-            let dist = (dx * dx + dy * dy + dz * dz).sqrt();
-            ej + bi.mass * bj.mass / dist
-        });
-        e + ei - ei2
-    })
+            e -= bi.mass * bj.mass / (dx * dx).sum().sqrt()
+        }
+    }
+    e
 }
 
-fn offset_momentum(bodies: &mut [Planet;N_BODIES]) {
-    let (px, py, pz) = bodies.iter()
-        .fold((0.0, 0.0, 0.0), |(px, py, pz), bi| 
-            (px + bi.vx * bi.mass,
-             py + bi.vy * bi.mass,
-             pz + bi.vz * bi.mass));
-
-    let sun = &mut bodies[0];
-    sun.vx = - px / SOLAR_MASS;
-    sun.vy = - py / SOLAR_MASS;
-    sun.vz = - pz / SOLAR_MASS;
-}
+pub fn advance(bodies: &mut [Body; N_BODIES], dt: f64) {
+    const N: usize = N_BODIES * (N_BODIES - 1) / 2;
 
-fn main() {
-    let n = std::env::args_os().nth(1)
-        .and_then(|s| s.into_string().ok())
-        .and_then(|n| n.parse().ok())
-        .unwrap_or(1000);
```

Comment:
If no argument are given, it must use 1000. That's used in the tests.

Reply:
Ok, I'll change this back.